New Read-only scheme

leerho opened the pull request:
This has a final boolean localReadOnly in each of BaseWritableBufferImpl
and BaseWritableMemoryImpl.

The resourceReadOnly in ResourceState still exists.  Now that this works, I can continue simplifying ResourceState. Let me know once you have reviewed and are OK with this.

All tests pass.

Commit:
New Read-only scheme

This has a final boolean localReadOnly in each of BaseWritableBufferImpl
and BaseWritableMemoryImpl.

All tests pass.

leventov commented on src/main/java/com/yahoo/memory/AllocateDirectMap.java:
```diff
@@ -60,6 +61,9 @@
    * @throws IOException file not found or RuntimeException, etc.
    */
   static AllocateDirectMap map(final ResourceState state) throws IOException {
+    if (isFileReadOnly(state.getFile())) {
```
If this is `AllocateDirectMap` (as opposed to `AllocateDirectWritableMap`, shouldn't it unconditionally do something like
```java
if (!state.isReadOnly()) {
  throw new IllegalStateException();
}
```
instead here?

leventov replied:
Or unconditionally set `state.setResourceReadOnly();`

leerho replied:
My goal is to make the resourceIsReadOnly_ flag a simple final (eliminating the StepBoolean), and I'm not there yet.  

What you see is correct in the current state of changes.  But in order to make this resource flag final this test and set needs to be moved earlier to when the ResourceState is first created.  This flag is strictly associated with the resource.  Each BaseWritable*Impl has its own RO flag set essentially by user choice of method calls.  The two flags are combined [here](https://github.com/DataSketches/memory/blob/NewReadOnlyScheme/src/main/java/com/yahoo/memory/WritableBuffer.java#L43) and [here](https://github.com/DataSketches/memory/blob/NewReadOnlyScheme/src/main/java/com/yahoo/memory/WritableMemory.java#L46).

leventov commented on src/main/java/com/yahoo/memory/AllocateDirectWritableMap.java:
```diff
@@ -38,8 +38,9 @@ private AllocateDirectWritableMap(final ResourceState state) {
-      throw new ReadOnlyException("File is read-only.");
+      throw new ReadOnlyException("Cannot map a read-only file into Writable Memory.");
     }
+    assert !state.isResourceReadOnly() : "state.resourceReadOnly should not be true";
```
Suggested check, not assert

leventov commented on src/main/java/com/yahoo/memory/BaseWritableMemoryImpl.java:
```diff
@@ -47,12 +47,15 @@
     capacity = state.getCapacity();
     cumBaseOffset = state.getCumBaseOffset();
+    this.localReadOnly = localReadOnly;
+
```
Extra line

leerho replied:
Thanks

leventov commented on src/main/java/com/yahoo/memory/WritableBufferImpl.java:
```diff
@@ -51,66 +51,82 @@
+    super(state, localReadOnly);
   }
 
+  //  static WritableBufferImpl newInstance(final ResourceState state, final boolean localReadOnly) {
```
garbage

leerho replied:
At first I thought I needed this, but I don't.  Will remove.

leventov commented on src/main/java/com/yahoo/memory/WritableBufferImpl.java:
```diff
@@ -51,66 +51,82 @@
   @Override
   public Memory asMemory() {
-    return asWritableMemory();
+    return asWritableMemoryImpl(true);
```
If on *writable* buffer `asMemory()` is called first, it will "install" read-only `originMemory`. And then, when `asWritableMemory()` is called, it will return not writable `WritableMemory`, that is wrong.

leerho replied:
You are right!  I think it needs to be changed to this:

```java
  private WritableMemory asWritableMemoryImpl(final boolean localReadOnly) {
    checkValid();
    if ((originMemory != null) && (originMemory.localReadOnly == localReadOnly)) {
      return originMemory;
    }
    return new WritableMemoryImpl(state, localReadOnly);
  }
```

In other words, the originMemory cannot be set from the Buffer side, only the Memory side.

leventov commented on src/main/java/com/yahoo/memory/WritableBufferImpl.java:
```diff
@@ -51,66 +51,82 @@
+    final WritableBufferImpl wBufImpl = new WritableBufferImpl(newState, localReadOnly);
     wBufImpl.setStartPositionEnd(0L, 0L, capacityBytes);
     return wBufImpl;
+
```
Extra line

leerho replied:
Thanks

Commit:
Review fixes

Commit:
resourceIsReadOnly_ flag is now final boolean.

Renamed ArrayOverflowTest to IgnoredArrayOverflowTest to make it easier
to find.

Step by Step :)
